openwisp_radius/base/models.py

```diff
@@ -1575,8 +1578,16 @@ def unverify_inactive_users(cls):
     def delete_inactive_users(cls):
         if not app_settings.DELETE_INACTIVE_USERS:
             return
+        cutoff_date = timezone.now() - timedelta(
+            days=app_settings.DELETE_INACTIVE_USERS
+        )
         User.objects.filter(
-            is_staff=False,
-            last_login__lt=timezone.now()
-            - timedelta(days=app_settings.DELETE_INACTIVE_USERS),
-        ).delete()
+            Q(is_staff=False)
+            & (
+                Q(last_login__lt=cutoff_date)
+                |
+                # There could be manually created users which never logged in.
+                # We use the "date_joined" field for such users.
+                Q(last_login__isnull=True, date_joined__lt=cutoff_date)
+            )
+        ).only('id', 'last_login', 'date_joined', 'is_staff').delete()
```

Comment:
BTW are we sure that adding `only` here does anything? This code should generate a `DELETE` query, not a `SELECT`, so I am not sure specifying the fields to select is useful at all here.
If it was useful we should add it also to the unverify method for consistency.

Reply:
Deleting a user performs these many queries 


```sql
(0.004) SELECT "openwisp_users_user"."last_login", "openwisp_users_user"."is_staff", "openwisp_users_user"."date_joined", "openwisp_users_user"."id" FROM "openwisp_users_user" WHERE (NOT "openwisp_users_user"."is_staff" AND ("openwisp_users_user"."last_login" < '2024-03-21T09:04:58.713418+00:00'::timestamptz OR ("openwisp_users_user"."date_joined" < '2024-03-21T09:04:58.713418+00:00'::timestamptz AND "openwisp_users_user"."last_login" IS NULL))); args=(datetime.datetime(2024, 3, 21, 9, 4, 58, 713418, tzinfo=datetime.timezone.utc), datetime.datetime(2024, 3, 21, 9, 4, 58, 713418, tzinfo=datetime.timezone.utc)); alias=default
(0.002) SELECT "account_emailaddress"."id" FROM "account_emailaddress" WHERE "account_emailaddress"."user_id" IN ('bfafed94-d6a3-4571-a447-87944aa1b028'::uuid); args=(UUID('bfafed94-d6a3-4571-a447-87944aa1b028'),); alias=default
(0.002) SELECT "socialaccount_socialaccount"."id" FROM "socialaccount_socialaccount" WHERE "socialaccount_socialaccount"."user_id" IN ('bfafed94-d6a3-4571-a447-87944aa1b028'::uuid); args=(UUID('bfafed94-d6a3-4571-a447-87944aa1b028'),); alias=default
(0.002) SELECT "radiustoken"."organization_id", "radiustoken"."created", "radiustoken"."key", "radiustoken"."user_id", "radiustoken"."can_auth" FROM "radiustoken" WHERE "radiustoken"."user_id" IN ('bfafed94-d6a3-4571-a447-87944aa1b028'::uuid); args=(UUID('bfafed94-d6a3-4571-a447-87944aa1b028'),); alias=default
(0.005) SELECT "openwisp_users_organizationuser"."created", "openwisp_users_organizationuser"."modified", "openwisp_users_organizationuser"."is_admin", "openwisp_users_organizationuser"."id", "openwisp_users_organizationuser"."user_id", "openwisp_users_organizationuser"."organization_id" FROM "openwisp_users_organizationuser" INNER JOIN "openwisp_users_organization" ON ("openwisp_users_organizationuser"."organization_id" = "openwisp_users_organization"."id") WHERE "openwisp_users_organizationuser"."user_id" IN ('bfafed94-d6a3-4571-a447-87944aa1b028'::uuid) ORDER BY "openwisp_users_organization"."name" ASC, "openwisp_users_organizationuser"."user_id" ASC; args=(UUID('bfafed94-d6a3-4571-a447-87944aa1b028'),); alias=default
(0.001) SELECT "openwisp_users_organizationowner"."created", "openwisp_users_organizationowner"."modified", "openwisp_users_organizationowner"."id", "openwisp_users_organizationowner"."organization_user_id", "openwisp_users_organizationowner"."organization_id" FROM "openwisp_users_organizationowner" WHERE "openwisp_users_organizationowner"."organization_user_id" IN ('6f7773c7-f53a-443b-b277-048a8b9eebb0'::uuid); args=(UUID('6f7773c7-f53a-443b-b277-048a8b9eebb0'),); alias=default
(0.000) BEGIN; args=None; alias=default
(0.001) DELETE FROM "authtoken_token" WHERE "authtoken_token"."user_id" IN ('bfafed94-d6a3-4571-a447-87944aa1b028'::uuid); args=(UUID('bfafed94-d6a3-4571-a447-87944aa1b028'),); alias=default
(0.001) DELETE FROM "radcheck" WHERE "radcheck"."user_id" IN ('bfafed94-d6a3-4571-a447-87944aa1b028'::uuid); args=(UUID('bfafed94-d6a3-4571-a447-87944aa1b028'),); alias=default
(0.001) DELETE FROM "radreply" WHERE "radreply"."user_id" IN ('bfafed94-d6a3-4571-a447-87944aa1b028'::uuid); args=(UUID('bfafed94-d6a3-4571-a447-87944aa1b028'),); alias=default
(0.002) DELETE FROM "radusergroup" WHERE "radusergroup"."user_id" IN ('bfafed94-d6a3-4571-a447-87944aa1b028'::uuid); args=(UUID('bfafed94-d6a3-4571-a447-87944aa1b028'),); alias=default
(0.001) DELETE FROM "radbatch_users" WHERE "radbatch_users"."user_id" IN ('bfafed94-d6a3-4571-a447-87944aa1b028'::uuid); args=(UUID('bfafed94-d6a3-4571-a447-87944aa1b028'),); alias=default
(0.001) DELETE FROM "openwisp_radius_phonetoken" WHERE "openwisp_radius_phonetoken"."user_id" IN ('bfafed94-d6a3-4571-a447-87944aa1b028'::uuid); args=(UUID('bfafed94-d6a3-4571-a447-87944aa1b028'),); alias=default
(0.001) DELETE FROM "openwisp_radius_registereduser" WHERE "openwisp_radius_registereduser"."user_id" IN ('bfafed94-d6a3-4571-a447-87944aa1b028'::uuid); args=(UUID('bfafed94-d6a3-4571-a447-87944aa1b028'),); alias=default
(0.001) DELETE FROM "openwisp_users_user_groups" WHERE "openwisp_users_user_groups"."user_id" IN ('bfafed94-d6a3-4571-a447-87944aa1b028'::uuid); args=(UUID('bfafed94-d6a3-4571-a447-87944aa1b028'),); alias=default
(0.001) DELETE FROM "openwisp_users_user_user_permissions" WHERE "openwisp_users_user_user_permissions"."user_id" IN ('bfafed94-d6a3-4571-a447-87944aa1b028'::uuid); args=(UUID('bfafed94-d6a3-4571-a447-87944aa1b028'),); alias=default
(0.002) DELETE FROM "openwisp_users_organizationinvitation" WHERE ("openwisp_users_organizationinvitation"."invited_by_id" IN ('bfafed94-d6a3-4571-a447-87944aa1b028'::uuid) OR "openwisp_users_organizationinvitation"."invitee_id" IN ('bfafed94-d6a3-4571-a447-87944aa1b028'::uuid)); args=(UUID('bfafed94-d6a3-4571-a447-87944aa1b028'), UUID('bfafed94-d6a3-4571-a447-87944aa1b028')); alias=default
(0.001) DELETE FROM "django_admin_log" WHERE "django_admin_log"."user_id" IN ('bfafed94-d6a3-4571-a447-87944aa1b028'::uuid); args=(UUID('bfafed94-d6a3-4571-a447-87944aa1b028'),); alias=default
(0.001) DELETE FROM "openwisp_users_organizationuser" WHERE "openwisp_users_organizationuser"."id" IN ('6f7773c7-f53a-443b-b277-048a8b9eebb0'::uuid); args=(UUID('6f7773c7-f53a-443b-b277-048a8b9eebb0'),); alias=default
(0.001) SELECT "openwisp_users_user"."password", "openwisp_users_user"."last_login", "openwisp_users_user"."is_superuser", "openwisp_users_user"."username", "openwisp_users_user"."first_name", "openwisp_users_user"."last_name", "openwisp_users_user"."is_staff", "openwisp_users_user"."is_active", "openwisp_users_user"."date_joined", "openwisp_users_user"."id", "openwisp_users_user"."email", "openwisp_users_user"."bio", "openwisp_users_user"."url", "openwisp_users_user"."company", "openwisp_users_user"."location", "openwisp_users_user"."phone_number", "openwisp_users_user"."birth_date", "openwisp_users_user"."notes", "openwisp_users_user"."language", "openwisp_users_user"."password_updated" FROM "openwisp_users_user" WHERE "openwisp_users_user"."id" = 'bfafed94-d6a3-4571-a447-87944aa1b028'::uuid LIMIT 21; args=(UUID('bfafed94-d6a3-4571-a447-87944aa1b028'),); alias=default
(0.001) SELECT "openwisp_users_organizationuser"."created", "openwisp_users_organizationuser"."modified", "openwisp_users_organizationuser"."is_admin", "openwisp_users_organizationuser"."id", "openwisp_users_organizationuser"."user_id", "openwisp_users_organizationuser"."organization_id", "openwisp_users_organization"."name", "openwisp_users_organization"."is_active", "openwisp_users_organization"."created", "openwisp_users_organization"."modified", "openwisp_users_organization"."slug", "openwisp_users_organization"."id", "openwisp_users_organization"."description", "openwisp_users_organization"."email", "openwisp_users_organization"."url", "openwisp_users_organizationowner"."created", "openwisp_users_organizationowner"."modified", "openwisp_users_organizationowner"."id", "openwisp_users_organizationowner"."organization_user_id", "openwisp_users_organizationowner"."organization_id" FROM "openwisp_users_organizationuser" INNER JOIN "openwisp_users_organization" ON ("openwisp_users_organizationuser"."organization_id" = "openwisp_users_organization"."id") LEFT OUTER JOIN "openwisp_users_organizationowner" ON ("openwisp_users_organizationuser"."id" = "openwisp_users_organizationowner"."organization_user_id") WHERE ("openwisp_users_organization"."is_active" AND "openwisp_users_organizationuser"."user_id" = 'bfafed94-d6a3-4571-a447-87944aa1b028'::uuid) ORDER BY "openwisp_users_organization"."name" ASC, "openwisp_users_organizationuser"."user_id" ASC; args=(UUID('bfafed94-d6a3-4571-a447-87944aa1b028'),); alias=default
(0.001) DELETE FROM "openwisp_users_user" WHERE "openwisp_users_user"."id" IN ('bfafed94-d6a3-4571-a447-87944aa1b028'::uuid); args=(UUID('bfafed94-d6a3-4571-a447-87944aa1b028'),); alias=default
(0.008) COMMIT; args=None; alias=default
```

Reply:
The update query is already pretty optimized. 

```sql 
UPDATE "openwisp_radius_registereduser" SET "is_verified" = false WHERE "openwisp_radius_registereduser"."user_id" IN (SELECT U0."user_id" FROM "openwisp_radius_registereduser" U0 INNER JOIN "openwisp_users_user" U1 ON (U0."user_id" = U1."id") WHERE (NOT (U0."method" IN ('', 'manual', 'email')) AND NOT U1."is_staff" AND U1."last_login" < '2024-03-21T09:53:52.283142+00:00'::timestamptz)); args=(False, '', 'manual', 'email', datetime.datetime(2024, 3, 21, 9, 53, 52, 283142, tzinfo=datetime.timezone.utc)); alias=default
```